File: Programmers/Lv2/triangle_snail.cpp
#include <bits/stdc++.h>

using namespace std;

enum snail_state { UP = 1, DOWN, RIGHT };

vector<string> input_func();
vector<int> solution(int n);

int main() {
    int n = 4;
    solution(n);
}

vector<string> input_func() {
    vector<string> input;

    string str, item;
    cin >> str;

    return input;
}
vector<int> solution(int n) {
    vector<int> answer;

    // init
    vector<vector<int>> snail(n);
    for (int i = 0; i < n; i++) {
        snail[i].resize(i + 1);
    }

    // run
    int col = 0, row = 0, state = DOWN, num = 1;
    for (int i = 0; i <= n; i++) {
        switch (state) {
            case UP:
                for (int j = i; j < n; j++) {
                    snail[col--][row--] = num++;
                }
                col += 2;
                row++;
                state = DOWN;
                break;

            case DOWN:
                for (int j = i; j < n; j++) {
                    snail[col++][row] = num++;
                }
                col--;
                row++;
                state = RIGHT;
                break;

            case RIGHT:
                for (int j = i; j < n; j++) {
                    snail[col][row++] = num++;
                }
                col--;
                row -= 2;
                state = UP;
                break;

            default:
                break;
        }
    }

    for (int i = 0; i < snail.size(); i++) {
        for (int j = 0; j < snail[i].size(); j++) {
            answer.push_back(snail[i][j]);
            cout << endl << answer.back();
        }
    }

    return answer;
}

File: java/src/programmers/lv2/triangle_snail.java
package programmers.lv2;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class triangle_snail {
    enum STATE {DOWN, RIGHT, UP}

    public int[] solution(int n) {
        int[] answer = new int[n *(n+1)/2];
        int[][] snail = new int[n][n];

        STATE state = STATE.DOWN;
        int x = 0, y = -1, cnt = 1;
        for (int i = 0; i <= n; i++) {
            for(int j=i; j < n; j++){
                if(state == STATE.DOWN) y++;
                else if(state == STATE.RIGHT) x++;
                else {y--;x--;}
                snail[y][x] = cnt++;
            }
            if(state == STATE.DOWN) state = STATE.RIGHT;
            else if(state == STATE.RIGHT) state = STATE.UP;
            else state = STATE.DOWN;
        }

        int k=0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (snail[i][j] == 0) break;
                answer[k++]=snail[i][j];
            }
        }
        return answer;
    }

    @Test
    void test() {
        Assertions.assertArrayEquals(new int[]{1, 2, 9, 3, 10, 8, 4, 5, 6, 7},solution(4));
    }
}
